packages/udt/src/udt/index.ts

```diff
@@ -274,33 +367,514 @@ export class Udt extends ssri.Trait {
+   * - Detecting minting operations (negative burned balance)
+   * - Ensuring sufficient UDT inputs are provided for transfers
+   */
+  async getBalanceBurned(
+    txLike: ccc.TransactionLike,
+    client: ccc.Client,
+  ): Promise<ccc.Num> {
+    const tx = ccc.Transaction.from(txLike);
+    return (
+      (await this.getInputsBalance(tx, client)) - this.getOutputsBalance(tx)
+    );
+  }
+
+  /**
+   * Completes UDT inputs for a transaction to cover the required output balance.
+   * This method automatically adds UDT inputs from the signer's available cells
+   * to ensure the transaction has sufficient UDT balance to cover all outputs,
+   * while also considering the capacity balance for transaction fees.
+   *
+   * @param txLike - The transaction to complete with UDT inputs
+   * @param from - The signer that will provide UDT inputs
+   * @param balanceTweak - Optional additional UDT balance to account for (e.g., for extra requirements)
+   * @returns A promise resolving to the number of inputs added to the transaction
+   *
+   * @throws {ErrorTransactionInsufficientCoin} When there are insufficient UDT cells to cover the required balance
+   *
+   * @example
+   * ```typescript
+   * const udt = new Udt(codeOutPoint, scriptConfig);
+   *
+   * // Create a transaction that needs 1000 UDT
+   * const tx = ccc.Transaction.from({
+   *   outputs: [
+   *     { lock: recipientLock, type: udt.script }
+   *   ],
+   *   outputsData: [ccc.numLeToBytes(1000, 16)]
+   * });
+   *
+   * // Add UDT inputs to cover the 1000 UDT requirement
+   * const addedInputs = await udt.completeInputs(tx, signer);
+   * console.log(`Added ${addedInputs} UDT inputs to the transaction`);
+   * ```
+   *
+   * @remarks
+   * This method implements smart input selection with dual balance consideration:
+   * - Calculates the required UDT balance (outputs + balanceTweak)
+   * - Considers both UDT balance and capacity balance for optimal selection
+   * - If existing inputs already cover the required UDT balance and capacity is non-negative, no new inputs are added
+   * - Adds inputs until both UDT balance requirement is met and capacity balance is non-negative
+   * - Uses accumulator pattern to track both UDT balance and capacity during input selection
+   * - Throws an error if the signer doesn't have sufficient UDT balance
+   */
+  async completeInputs(
+    txLike: ccc.TransactionLike,
+    from: ccc.Signer,
+    balanceTweak?: ccc.NumLike,
+    capacityTweak?: ccc.NumLike,
+  ): Promise<number> {
+    const tx = ccc.Transaction.from(txLike);
+    const balanceBurned =
+      (await this.getBalanceBurned(tx, from.client)) -
+      ccc.numFrom(balanceTweak ?? 0);
+    const capacityBurned =
+      (await tx.getFee(from.client)) - ccc.numFrom(capacityTweak ?? 0);
+
+    if (balanceBurned >= ccc.Zero && capacityBurned >= ccc.Zero) {
+      return 0;
+    }
+
+    const { addedCount, accumulated } = await tx.completeInputs(
```

Comment:
This is difficult to override, it could make sense to partition this in multiple overridable methods

Reply:
For example it could make sense to have a generator methods that returns said cells wrapped, possibly wrapped into an object where each cell is annotated with its ckb and udt value

Reply:
Yeah, you're right, my underlying issue is with `tx.completeInputs`: it may be a fine piece of engineering, I don't like it, my bad!! :rofl: